WMF Framework/Event Platform/StorageManager.swift

```diff
@@ -0,0 +1,185 @@
+import Foundation
+import CocoaLumberjackSwift
+
+class StorageManager: NSObject {
+
+    private var managedObjectContext: NSManagedObjectContext
+    private let pruningAge: TimeInterval = 60*60*24*30 // 30 days
```

Comment:
Did we pick a magic number here? Should we do some napkin calculations and make sure this makes sense?

Reply:
I checked the commit adding this (https://github.com/wikimedia/wikipedia-ios/commit/9fe6d9389a3cfe059f9747bfc826e25ff4d00384) and its enclosing PR (https://github.com/wikimedia/wikipedia-ios/pull/1484) and I can't find any specific justification for a 30-day window.  The relevant Phab task indicates that technical planning was being done in Google Docs, but the one I was able to find in my email has been deleted. :(

There is a wiki page here with useful discussion about what they were trying to accomplish, but nothing directly on point: https://www.mediawiki.org/wiki/Wikimedia_Apps/App_Analytics

Reply:
One thing that occurs to me is that if we're trying to gauge the effect on event numbers of putting DB persistence back in, it would be best to match the old client on implementation details like this one as a control.  

In the long run, it's probably worth creating a Phab task to hash it out with PM input and to document the decision for posterity.

Reply:
Thanks for doing all of that digging! Now that I'm pulling the sweater thread a little bit, I still think we aren't quite all on the same page about how our various requirements are interacting, though. 

The output buffer fulfills two basic purposes:
1. It allows events to pool so we can apply bursting and batching strategies.
2. It allows events to survive transient network connectivity loss

The key word here is 'transient', since events will only find their way to the output buffer for reason #2 if they occur
- DURING a loss of network connectivity
- AFTER stream configuration load
- BEFORE app termination

This is because stream config load requires a network connection. That means that the actual TTL is likely to be the time until the user terminates the app for the first time while offline. At that point, no further events will ever make their way to the output buffer until a new stream configuration is fetched.

This behavior is intentional -- both the stream configuration and the input buffer should not be persisted to disk, and therefore should not persist beyond app termination. That means that after app termination, no further events will ever make their way to the output buffer until a new stream configuration is fetched.

My basic point here is that the true TTL is `min(30 days, time to first app termination after network loss)`. I think my suggestion would be to set a size limit rather than a TTL, since there is already an implicit TTL (the time to app termination). This would also protect users against any bugs which could cause an instrument to spray a lot of events and start filling up storage. Certainly after 30 days without phoning home, it would be possible to get into a bad situation.

Reply:
> This behavior is intentional -- both the stream configuration and the input buffer should not be persisted to disk, and therefore should not persist beyond app termination.

@linehan It seems like this will still drop many events if we are only persisting the output buffer. In my mind input buffer events that are made while offline without a stream configuration are valid and should be given a chance to post between app terminations. But I admit I am clouded by the old client library design.

I'm fine with having a pruning be a size limit of the db events table rather than a particular date - just don't understand the reasons for the input / output buffer distinction now that we have persistence. This PR does still have the config analysis / inSample mapping of course, it just occurs later right before posting rather than coordinating between the two separate buffers.

Reply:
> > This behavior is intentional -- both the stream configuration and the input buffer should not be persisted to disk, and therefore should not persist beyond app termination.
>
> In my mind input buffer events that are made while offline without a stream configuration are valid and should be given a chance to post between app terminations.

@tonisevener I've drifted a bit here, myself, but the reason we initially included an input buffer was to cover a "blind spot" between app launch and stream config fetch. This is a pattern from MediaWiki that has proven useful. It wasn't intended as a general way to "delay judgement" on events in the absence of stream configuration, except in that small window between load and fetch. We should talk more about this, but here are a couple of properties that result from minimizing the input buffer's role:
- The system will "fail safe" in the absence of stream config. If there is a bug or problem with any instrument or stream, it can be either shut off in stream config; if a device cannot fetch steam config, it will be disabled by default (along with all streams and instruments)
- We can be sure that the budget for events/storage/performance on a given device is an exact reflection of the stream configuration. If input buffer is optimistic about the possibility of one day getting a stream config, our budgets will need to instead reflect the possibility that, no matter what our stream config is, all events from all instruments and all streams will fire and be stored to disk for the TTL.

The summary is that it makes the stream configuration a better reflection of the conditions on a device, which helps us reason about and control various factors better.

Reply:
@linehan Makes sense. I can see the argument both ways. If today's stream config is supposed to dictate all past and present events (i.e. the current config sampling rate of zero should shut down events logged 2 weeks ago), we should scrap the input buffer. Or if you need events posted to represent what the config settings were at the time the event was generated, we should keep the input buffer.

My main concern was that by keeping the input buffer, persistence only catches those whose stream config fetch succeeds but posting fails and the app is terminated (like you said, transient network connection issues, not full offline mode). I would guess that number is far lower than situations where both stream config and posting fails.

Reply:
I guess I was assuming that we wanted unsent event persistence across app termination and relaunch simply by virtue of the fact that we're discussing DB persistence.  Otherwise, why take on the complexity?  Is there any point?

That said, I think that dropping unsent events at app termination is the right thing to do, and 30 days seems quite excessive as a max TTL.

Reply:
OK, I can see a justification if we expect the amount of pending event data between bursts and/or transient losses of network connectivity to become larger than we want to hold in memory at a given time. Although if that's the case, we're arguably creating an excessive number of events...

Reply:
> My main concern was that by keeping the input buffer, persistence only catches those whose stream config fetch succeeds but posting fails and the app is terminated (like you said, transient network connection issues, not full offline mode). I would guess that number is far lower than situations where both stream config and posting fails.

Yes, so we agree persistence in this situation is more confusing than useful, *if* the other behavior is what I described, I hear what you're saying though, that offline mode is a thing. I don't know how much of a thing it is, but let's say it's enough of a thing.

I'll lay out the full three approaches (at least as far as I can formulate them) rather than just talking about one or the other.

Approach 1 - "The future is the present"
- Don't persist the stream configuration
- Persist the input buffer
- Input buffer events are processed according to the stream configuration that is eventually fetched, regardless of when they occurred.
  - This leads to events in seeming to have been generated prior to their stream configuration allowing it, or with properties out of step with their stream configuration
  - It's possible to solve this, but it's not solved out-of-the-box today

Approach 2 - "The past is the present"
- Persist the most recent stream configuration
- Don't persist the input buffer
- If stream configuration fetch fails, use the persisted one instead; events are processed through to the output buffer.
   - This leads to the same situation in which events may seem to have been generated subsequent to their stream configuration disallowing it, or with properties out of step with their stream configuration.
      - It's possible to solve this, in a way that is more likely to get implemented soon (the stream configurations have a version or modtime, and events get rejected at the server if their event doesn't give the current version/modtime).
   - The other problem here is that this will not "fail safe" if the stream configuration change disables a buggy instrument or stream

Approach 3 - "The present is the present"
- Don't persist the stream configuration
- Don't persist the input buffer
- If stream configuration fetch fails, no events get sent.
  - Events will not be recorded for users that have not fetched a stream configuration. Offline mode, etc. gets hard to instrument.
    - Possible to solve this by having different behavior for 'offline mode' events, or by caching some stream configuration, etc. 
    - Seems easier to evolve from this to a more sophisticated solution (at least to me)

Reply:
OK, I put the input buffer back in to implement Approach 3.